### PR TITLE
Implement autologging (tracing-only) for llama-index

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/12677?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12677/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12677
```

</p>
</details>

### What changes are proposed in this pull request?

Implement barebone autologging support (only tracing) for llama-index. Since tracer (i.e. span/event handlers) is managed via global variable in llama index, we don't need to patch or inject function calls.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Testing the tracing functionality in staging.
Screenshot 2024-07-16 at 15 21 40

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

Will file a PR for the main documentation page today.

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/llama_index/__init__.py
Comment (BenWilson2) [inline mlflow/llama_index/__init__.py:359]: +1

File: mlflow/llama_index/tracer.py
Comment (BenWilson2) [inline mlflow/llama_index/tracer.py:50]: Is the class name check here solely to avoid a local scoped import to avoid a circular dep reference?

**Review by michael-berk — APPROVED** [review]
LGTM (but should get another approval)

File: mlflow/llama_index/__init__.py
Comment (michael-berk) [inline mlflow/llama_index/__init__.py:343]: Nit: should we modify this to be `log_traces`? 

If a user has code running with `autolog(log_traces=True)`, as MLflow updates the functionality the user code functionality will not change. With `disable`, behavior will slowly change over time

Comment (BenWilson2) [reply mlflow/llama_index/__init__.py:343]: +1 let's keep a consistent interface so that when we have support for the remaining items (log_models, log_model_signatures, log_input_examples) we can retain that common interface.

**Review by BenWilson2 — APPROVED** [review]
LGTM once the trace argument interface is unified with other autologging implementations

File: mlflow/llama_index/tracer.py
Comment (B-Step62) [reply mlflow/llama_index/tracer.py:50]: oh it doesn't have to be name check, will replace with `isinstance`

File: mlflow/llama_index/__init__.py
Comment (B-Step62) [reply mlflow/llama_index/__init__.py:343]: makes sense!

**Pull request merged**
